rfcs/react-components/convergence/reusing-avatar-in-components.md

```diff
@@ -0,0 +1,259 @@
+  // recomposed TableAvatar can read TableContext to adapt to its size prop
+  const { size } = useTableContext();
+
+   const state = useAvatar({...props, size: tableAvatarSizeMap[size]});
+   state = useAvatarStyles(props);
+   return renderAvatar(state);
+}
+```
+
+```tsx
+<Table size="smaller">
+  <TableBody>
+    <TableRow>
+      <TableCell>
+        {/* This will be sized automtically */}
+        <TableCellLayout media={<TableAvatar />}>Main content</TableCellLayout>
+      </TabeCell>
+    </TableRow>
+  </TableBody>
+</Table>
+```
+
+#### Pros
+
+- Best level of support for component reuse
+- Can control everything from style override to props
+
+#### Cons
+
+- Lots of variations of a base component, TableAvatar, PersonaAvatar, ToolarAvatar...
+- How do users know they should use a 'special' variant of the component?
+
+### Separate slots
+
+This solution is what the `Alert` currently does (however, it does not seem necessary in that component):
+
+```tsx
+export function useTableCellLayout() {
+  const state = {
+    components: {
+      icon: 'span',
+      avatar: Avatar,
+    }
+    avatar: resolveShorthand(props.avatar);
+  }
+
+  /** Avatar prop takes precedence over the icon*/
+  if (!avatar) {
+    icon = resolveShorthand(props.icon, {
+      defaultProps: {
+        children: defaultIcon,
+      },
+    });
+    state.icon = icon;
+  }
+
+  return state;
+}
+
+export function renderTableCellLayout() {
+
+}
+```
+
+```tsx
+<Table size="smaller">
+  <TableBody>
+    <TableRow>
+      <TableCell>
+        <TableCellLayout icon={<Icon />}>Main content</TableCellLayout>
+      </TabeCell>
+      <TableCell>
+        {/* Internally the slot will override the size */}
+        <TableCellLayout avatar={{ presence: 'available' }}>Main content</TableCellLayout>
+      </TabeCell>
+    </TableRow>
+  </TableBody>
+</Table>
+```
+
+#### Pros
+
+- Very obvious to know what 'media' components are supported
+- Can configure the component through our well defined (and typesafe) shorthand API
+
+#### Cons
+
+- Bundle size - users who don't want to use avatar will have it in their bundle
+- Confusing priorities - One slot wins over the other, it's why we have `appearance` prop
+
+### Shared context
+
+We could consider a `AvatarContext` that is a part of `react-shared-contexts` that can be used by components that wish
+to reuse the `Avatar`.
+
+```tsx
+const tableAvatarSizeMap = {
+  small: 24,
+  smaller: 20,
+};
+
+export function renderTable(state) {
+  const { size } = state;
+  const { slots, slotProps } = getSlots(state);
+
+  return (
+    <slots.root {...slotProps.root}>
+      <AvatarContext size={tableAvatarSizeMap[size]}>{slotProps.root.children}</AvatarContext>
+    </slots.root>
+  );
+}
+```
+
+#### Pros
+
+- We also use context for lots of other user cases - it works
+
+#### Cons
+
+- Extra code in base component
+- Needs prototyping and investigation
+- All avatars under this context will be affected
```

Comment:
I suggest to add example with `Popover` there - an `Avatar` inside a `Popover` will have size "20" instead of default one

Reply:
As I implemented in the prototype, the choice of slot to apply this context can be more 'intelligent' https://github.com/microsoft/fluentui/pull/24807